rafthttp/snapshot.go

```diff
@@ -0,0 +1,188 @@
+	snapCloseTimer *time.Timer
+}
+
+// addSnap adds snapshot into snapshot buffer.
+// The added snapshot waits to be written out through writeSnapTo func.
+// If the snapshot exceeds wait timeout, it is closed and snapshot buffer removes it.
+// At most one snapshot could be in the buffer at the same time, or it panics.
+func (s *snapshotBuffer) addSnap(to types.ID, index uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.hasSnap {
+		plog.Panicf("unexpected add snapshot when previous one has not finished")
+	}
+	s.hasSnap = true
+	s.snapTo = to
+	s.snapIndex = index
+	s.snapCloseTimer = time.AfterFunc(snapshotWaitTimeout, func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if !s.hasSnap {
+			return
+		}
+		s.hasSnap = false
+		// close the snapshot in the store
+		s.snapst.WriteSnapTo(&errWriter{errors.New("cancelled")})
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out due to timeout", index, s.snapTo)
+		s.r.ReportSnapshot(uint64(to), raft.SnapshotFailure)
+	})
+}
+
+// writeSnapTo writes the snapshot with the given index into the given writer.
+func (s *snapshotBuffer) writeSnapTo(w io.Writer, index uint64) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.hasSnap {
+		return fmt.Errorf("failed to find available snapshot")
+	}
+	if s.snapIndex != index {
+		return fmt.Errorf("failed to find the snapshot with index %d", index)
+	}
+	s.hasSnap = false
+	s.snapCloseTimer.Stop()
+
+	err := s.snapst.WriteSnapTo(w)
+	if err != nil {
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", index, s.snapTo, err)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFailure)
+		return err
+	}
+	plog.Infof("snapshot [index: %d, to: %s] sent out successfully", index, s.snapTo)
+	s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFinish)
+	return nil
+}
+
+// snapshotProcessor processes incoming raft snapshot message.
+type snapshotProcessor struct {
+	tr       http.RoundTripper
+	upGetter urlPickerGetter
+	r        Raft
+	snapst   SnapshotStore
+}
+
+// processSnap processes the given raft snapshot message.
+// First, it sends requests to get full snapshot for raft snapshot message. Then
+// it lets snapshot store read out the snapshot. After the full snapshot is
+// received, it lets raft process the raft snapshot message.
+func (s *snapshotProcessor) processSnap(m raftpb.Message) {
+	remote := types.ID(m.From)
+	index := m.Snapshot.Metadata.Index
+	snapr, err := s.getSnapFrom(remote, index)
+	if err != nil {
+		plog.Warningf("failed to request snapshot [index: %d] (%v)", m.Snapshot.Metadata.Index, err)
+		return
+	}
+	defer snapr.Close()
+
+	if err := s.snapst.SaveSnapFrom(snapr, index); err != nil {
+		plog.Warningf("failed to read snapshot [index: %d] (%v)", m.Snapshot.Metadata.Index, err)
+		return
+	}
+	if err := s.r.Process(context.TODO(), m); err != nil {
+		plog.Warningf("failed to process raft message (%v)", err)
+	}
+}
+
+// getSnapFrom gets full snapshot with the given index from the remote member.
```

Comment:
// getSnapFrom requests the snapshot data with given index from the remote.
// It returns a ReadCloser for the snapshot data when succeeds. Otherwise, an
// error will be returned.